criu/memfd.c

```diff
@@ -235,20 +227,58 @@ static struct memfd_inode *memfd_alloc_inode(int id)
 	return inode;
 }
 
-extern int restore_memfd_shm(int fd, u64 id, u64 size);
+static LIST_HEAD(inode_list);
+struct memfd_inode_node {
+	MemfdInodeEntry *mie;
+	struct list_head node;
+};
+
+static int collect_one_memfd_inode(void *o, ProtobufCMessage *base, struct cr_img *i)
+{
+	MemfdInodeEntry *mie = pb_msg(base, MemfdInodeEntry);
+	struct memfd_inode_node *n = o;
+
+	n->mie = mie;
+
+	list_add_tail(&n->node, &inode_list);
+
+	return 0;
+}
+
+static struct collect_image_info memfd_inode_cinfo = {
+	.fd_type = CR_FD_MEMFD_INODE,
+	.pb_type = PB_MEMFD_INODE,
+	.priv_size = sizeof(struct memfd_inode_node),
+	.collect = collect_one_memfd_inode,
+	.flags = COLLECT_NOFREE,
```

Comment:
nit: it's not so important (CRIU works for a small amount of time), but as far as I understand from implementation https://github.com/checkpoint-restore/criu/blob/8cf5f298ebefe49e2977769039849348dc199e9f/criu/protobuf.c#L244
`COLLECT_NOFREE` implies not free memory after collecting object from pb image. As I see from code we using pointers to `MemfdInodeEntry`, so, we need to have a valid pointer for a long time. But when we finished with memfd restoration shouldn't we call `memfd_inode_entry__free_unpacked` somewhere? It looks like we currently not have the general infrastructure of "destructors" for that.

Reply:
We don't want to spend time to release these objects because from the moment when we know that we don't need them to the moment when the criu process will die is just one step of resuming processes.